server/services/store/sqlstore/blocks.go

```diff
@@ -516,6 +516,76 @@ func (s *SQLStore) deleteBlock(db sq.BaseRunner, c store.Container, blockID stri
 
 	return nil
 }
+
+func (s *SQLStore) undeleteBlock(db sq.BaseRunner, c store.Container, blockID string, modifiedBy string) error {
+	blocks, err := s.getBlockHistory(db, c, blockID, model.QueryBlockHistoryOptions{Limit: 1, Descending: true})
+	if err != nil {
+		return err
+	}
+
+	if len(blocks) == 0 {
+		return nil // deleting non-exiting block is not considered an error (for now)
+	}
+	block := blocks[0]
+
+	if block.DeleteAt == 0 {
+		return nil // undeleting not deleted block is not considered an error (for now)
+	}
+
+	fieldsJSON, err := json.Marshal(block.Fields)
+	if err != nil {
+		return err
+	}
+
+	now := utils.GetMillis()
+	columns := []string{
+		"workspace_id",
+		"id",
+		"parent_id",
+		s.escapeField("schema"),
+		"type",
+		"title",
+		"fields",
+		"root_id",
+		"modified_by",
+		"create_at",
+		"update_at",
+		"delete_at",
+		"created_by",
+	}
+
+	values := []interface{}{
+		c.WorkspaceID,
+		block.ID,
+		block.ParentID,
+		block.Schema,
+		block.Type,
+		block.Title,
+		fieldsJSON,
+		block.RootID,
+		modifiedBy,
+		block.CreateAt,
+		now,
+		0,
+		block.CreatedBy,
+	}
+	insertHistoryQuery := s.getQueryBuilder(db).Insert(s.tablePrefix + "blocks_history").
+		Columns(append(columns, "insert_at")...).
+		Values(append(values, now)...)
```

Comment:
Should this be deleted from the history table. Otherwise isn't the history only effectively one level deep. Perhaps that is the desire.

Reply:
I think that is the desire. Just adding a new event in the history, it was deleted, and just right after was no longer deleted. The history keep consistent. Even if you want something like "restore" old version, for me is exactly the same, you take the status that was before, and create a new history entry with that data.